### PR TITLE
Add a fix and test for setting an empty string as a key in a map

### PR DESCRIPTION
**Pull request by jonludlam**
Also fix other tests that already rely on this working

Signed-off-by: Jon Ludlam <jonathan.ludlam@citrix.com>

### CODE REVIEW

File: ocaml/database/db_cache_types.ml [hunk 1]
```diff
@@ -351,6 +351,7 @@ let remove_from_set key t =
 
 exception Duplicate
 let add_to_map ~idempotent key value t =
+  if String.length key = 0 then raise Db_exn.Empty_key_in_map;
```
Comment (lindig) [inline ocaml/database/db_cache_types.ml]: This is more expensive than necessary because it always takes the string length. It's cheaper to test `if key = "" then .. `

File: ocaml/database/string_marshall_helper.ml [hunk 2]
```diff
@@ -28,11 +28,14 @@ let ensure_utf8_xml string =
     warn "Whilst doing 'set' of structured field, string truncated to: '%s'." prefix;
   prefix
 
+let ensure_nonempty s =
+  if String.length s = 0 then raise Db_exn.Empty_key_in_map else s
```
Comment (lindig) [inline ocaml/database/string_marshall_helper.ml]: Same as above

File: ocaml/database/db_cache_types.ml (hunk 1 above)
Comment (jonludlam) [reply ocaml/database/db_cache_types.ml]: https://github.com/ocaml/ocaml/blob/4.04/byterun/str.c#L39-L45

Comment (jonludlam) [reply ocaml/database/db_cache_types.ml]: I'm not sure which is cheaper to be honest.

Comment (lindig) [reply ocaml/database/db_cache_types.ml]: It's fine as it is because this operation is cheap, as your link to the C is showing.
